### ISSUE TITLE
Sqlserver datetime2 precision not respected in batch insert

### ISSUE THREAD

**Issue by JWood48** [jOOQ/jOOQ#10093]
I have a JOOQ setup where the tables are generated based on liquibase and a H2 database

When created in Sqlserver it uses datetime2 type for precision. The table looks like this:
```sql
CREATE TABLE [eventstore].[TimestampTest](
	[evt_id] [bigint] IDENTITY(1,1) NOT NULL,
	[timestamp] [datetime2](7) NOT NULL,
 CONSTRAINT [PK_TimestampTestId] PRIMARY KEY CLUSTERED 
(
	[evt_id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY]
) ON [PRIMARY]
```

When inserting new rows using direct insertion and record batch insertion the datetim2 precision is ok (even though they are cast to datetime), when using batch insertion the precision is not ok!

TEST: 

```java
Timestamp roundingIssueTimestamp = Timestamp.valueOf("2020-04-16 16:53:28.467");

log.info("INITIAL_TIMESTAMP: "+roundingIssueTimestamp);

log.info("DIRECT_INSERTION");
ctx.insertInto(Timestamptest.TIMESTAMPTEST)
.set(Timestamptest.TIMESTAMPTEST.TIMESTAMP, roundingIssueTimestamp)
.execute();
;

log.info("BATCH_INSERTION");
ctx.batch(
	ctx
	.insertInto(Timestamptest.TIMESTAMPTEST)
	.set(Timestamptest.TIMESTAMPTEST.TIMESTAMP, roundingIssueTimestamp)
	).execute(); 		


log.info("RECORD_BATCH_INSERTION");

Record r1 = Timestamptest.TIMESTAMPTEST.newRecord();
r1.set(Timestamptest.TIMESTAMPTEST.TIMESTAMP, roundingIssueTimestamp);

Record r2 = Timestamptest.TIMESTAMPTEST.newRecord();
r2.set(Timestamptest.TIMESTAMPTEST.TIMESTAMP, roundingIssueTimestamp);

ctx.insertInto(Timestamptest.TIMESTAMPTEST)
.set(r1)
.newRecord()
.set(r2)
.execute();

log.info("FETCH_ALL");
Result<Record> fetch = ctx.selectFrom(Timestamptest.TIMESTAMPTEST).fetch();
```

4 rows are inserted all with the same timestamp!

One is directly inserted
One is inserted in a batch statement
Two are inserted as records in a batch

Expected result would be 4 rows in the db all with the same timestamp, but the batch inserted row seems to be rounded to datetime type before being inserted in the datetime2 field?
 

OUTPUT:
```logback
MSSQLMigrationTest - INITIAL_TIMESTAMP: 2020-04-16 16:53:28.467
MSSQLMigrationTest - DIRECT_INSERTION
LoggerListener  - Executing query          : insert into [robex_eventstore].[TimestampTest] ([timestamp]) values (?)
LoggerListener  - -> with bind values      : insert into [robex_eventstore].[TimestampTest] ([timestamp]) values (cast('2020-04-16 16:53:28.467' as datetime))
LoggerListener  - Affected row(s)          : 1
MSSQLMigrationTest - BATCH_INSERTION
BatchSingle     - Single batch             : No bind variables have been provided with a single statement batch execution. This may be due to accidental API misuse
LoggerListener  - Executing batch query    : insert into [robex_eventstore].[TimestampTest] ([timestamp]) values (cast('2020-04-16 16:53:28.467' as datetime))
MSSQLMigrationTest - RECORD_BATCH_INSERTION
LoggerListener  - Executing query          : insert into [robex_eventstore].[TimestampTest] ([timestamp]) values (?), (?)
LoggerListener  - -> with bind values      : insert into [robex_eventstore].[TimestampTest] ([timestamp]) values (cast('2020-04-16 16:53:28.467' as datetime)), (cast('2020-04-16 16:53:28.467' as datetime))
LoggerListener  - Affected row(s)          : 2
MSSQLMigrationTest - FETCH_ALL
LoggerListener  - Executing query          : select [robex_eventstore].[TimestampTest].[evt_id], [robex_eventstore].[TimestampTest].[timestamp] from [robex_eventstore].[TimestampTest]
LoggerListener  - Fetched result           : +------+---------------------------+
LoggerListener  -                          : |evt_id|timestamp                  |
LoggerListener  -                          : +------+---------------------------+
LoggerListener  -                          : |     1|2020-04-16 16:53:28.467    |
LoggerListener  -                          : |     2|2020-04-16 16:53:28.4666667|
LoggerListener  -                          : |     3|2020-04-16 16:53:28.467    |
LoggerListener  -                          : |     4|2020-04-16 16:53:28.467    |
LoggerListener  -                          : +------+---------------------------+
LoggerListener  - Fetched row(s)           : 4
```

The second row with id 2 is not the same as the other rows!!

### Versions:

- jOOQ: 3.11.11 pro
- Java: 1.8
- Database (include vendor): sqlserver + H2
- OS: Windows
- JDBC Driver (include name if inofficial driver): com.microsoft.sqlserver.jdbc.SQLServerDriver (8.2.2.jre8)

**Comment by lukaseder** [comment]
Thanks a lot for your report. I'll try to reproduce this in a bit and will revert back to you ASAP

**Comment by lukaseder** [comment]
Thank you very much for your patience. I could reproduce this also on jOOQ 3.14.0-SNAPSHOT. Getting the same inconsistent result:

```
+----+---------------------------+
|  id|ts                         |
+----+---------------------------+
|   1|2020-04-16 16:53:28.467    |
|   2|2020-04-16 16:53:28.4666667|
|   3|2020-04-16 16:53:28.467    |
|   4|2020-04-16 16:53:28.467    |
+----+---------------------------+
```

Will investigate

**Comment by lukaseder** [comment]
Just to be sure, I checked a plain JDBC example, and that dose not expose the difference:

```java
String sql = "insert into t_10093 (id, ts) values (?, ?)";

try (
    PreparedStatement s1 = c.prepareStatement(sql);
    PreparedStatement s2 = c.prepareStatement(sql);
) {
    s1.setInt(1, 1);
    s1.setTimestamp(2, ts);
    s1.executeUpdate();

    s2.setInt(1, 2);
    s2.setTimestamp(2, ts);
    s2.addBatch();
    s2.executeBatch();

    try (Statement s = c.createStatement(); ResultSet rs = s.executeQuery("select * from t_10093")) {
        while (rs.next()) {
            System.out.println("ID: " + rs.getInt(1) + ", TS: " + rs.getTimestamp(2));
        }
    }
}
```

This produces:

```
ID: 1, TS: 2020-04-16 16:53:28.467
ID: 2, TS: 2020-04-16 16:53:28.467
```

**Comment by lukaseder** [comment]
Notice the log message:

```
BatchSingle     - Single batch             : No bind variables have been provided with a single statement batch execution. This may be due to accidental API misuse
```

There are two overloads `DSLContext.batch(Query)` (normally for use with bind variables) and `DSLContext.batch(Query...)` (for use without bind variables). The first one is often called by accident.

**Comment by lukaseder** [comment]
Clearly, jOOQ's current behaviour is dated. We shouldn't default to using `DATETIME` throughout jOOQ, where a `SQLDataType.TIMESTAMP` is involved. I'll fix this right away for jOOQ 3.14: https://github.com/jOOQ/jOOQ/issues/10146

Prior to upgrading, you should be able to change jOOQ's behaviour by specifying a custom data type binding: https://www.jooq.org/doc/latest/manual/sql-building/queryparts/custom-bindings

In that binding, you can override the cast expression that is produced when inlining the bind value, e.g. for batch queries. Let me know if you need any help with that and I'll be very happy to assist you.

**Comment by lukaseder** [comment]
#10146 fixes this issue here, as the cast to `datetime2` produces the expected value.

Thanks again for your report!
